h_malloc.c

```diff
@@ -694,7 +786,17 @@ static inline void deallocate_small(void *p, const size_t *expected_size) {
     if (likely(!is_zero_size)) {
         check_canary(metadata, p, size);
 
-        if (ZERO_ON_FREE) {
+        bool skip_zero = false;
+#ifdef HAS_ARM_MTE
+        if (likely(is_memtag_enabled())) {
+            arm_mte_store_tags_and_clear(set_pointer_tag(p, RESERVED_TAG), size);
+            // metadata->arm_mte_tags is intentionally not updated, it should keep the previous slot
+            // tag after slot is freed
```

Comment:
```suggestion
            // metadata->arm_mte_tags is intentionally not updated, since a freed slot
            // should keep its tag.
```

Reply:
This variant sounds confusing to me (freed slots are tagged with a reserved tag), though I agree that the original needs improvement.

Reply:
Referenced tag_and_clear_slab_slot() instead: https://github.com/GrapheneOS/hardened_malloc/pull/225/commits/99ab17364e71f8eadbc6adda00db88e9733fd537